components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/internal/DefaultServiceURLBuilder.java

```diff
@@ -0,0 +1,188 @@
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addParameter(String key, String value) {
+
+        parameters = new HashMap<>();
+        parameters.put(key, value);
+        return this;
+    }
+
+    /**
+     * This method is called when the URL needs to be appended with a fragment.
+     *
+     * @param fragment Fragment.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder setFragment(String fragment) {
+
+        this.fragment = fragment;
+        return this;
+    }
+
+    /**
+     * This method is called when the URL fragment needs to be appended with parameters. Such parameters should be
+     * entered as <k,v> pairs. These parameters will get appended with an "&".
+     *
+     * @param key   Key.
+     * @param value Value.
+     * @return {@link ServiceURLBuilder}.
+     */
+    @Override
+    public ServiceURLBuilder addFragmentAsParams(String key, String value) {
+
+        fragmentParams = new HashMap<>();
+        fragmentParams.put(key, value);
+        return this;
+    }
+
+    @Override
+    public ServiceURL build() throws URLBuilderException {
+
+        String protocol = fetchProtocol();
+        String hostName = fetchHostName();
+        int port = fetchPort();
+        String resolvedUrlContext = resolveUrlPath(urlPaths);
+        String tenantDomain = resolveTenantDomain();
+
+        StringBuilder resolvedUrlStringBuilder = new StringBuilder();
+        if (isTenantQualifiedUrlsEnabled()) {
+            if (StringUtils.isNotBlank(tenantDomain)) {
+                resolvedUrlStringBuilder.append("/t/").append(tenantDomain);
+            }
+        }
+
+        if (StringUtils.isNotBlank(resolvedUrlContext)) {
+            if (resolvedUrlContext.trim().charAt(0) != '/') {
+                resolvedUrlStringBuilder.append("/").append(resolvedUrlContext.trim());
+            } else {
+                resolvedUrlStringBuilder.append(resolvedUrlContext.trim());
+            }
+        }
+        return new ServiceURL(protocol, hostName, port, resolvedUrlStringBuilder.toString(), parameters, fragment,
+                fragmentParams);
+    }
+
+    private String resolveTenantDomain() {
+
+        String tenantDomain = IdentityTenantUtil.getTenantDomainFromContext();
+        if (StringUtils.isBlank(tenantDomain)) {
+            tenantDomain = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+        }
+        return tenantDomain;
+    }
+
+    private String resolveUrlPath(String[] urlPaths) {
+
+        StringBuilder urlPathBuilder = new StringBuilder();
+        if (ArrayUtils.isNotEmpty(urlPaths)) {
+            for (String path : urlPaths) {
+                urlPathBuilder.append(path).append("/");
+            }
+            if (urlPathBuilder.toString().endsWith("/")) {
+                urlPathBuilder.setLength(urlPathBuilder.length() - 1);
```

Comment:
Is this the best way you can remove '/' from end. I think you can use charAt

Reply:
Looked into this and found that setting the length in the above manner is more efficient than using `stringbuilder.deleteCharAt(int index)`

ref: []()https://stackoverflow.com/a/10131077/6113775

Reply:
mmm but you are creating a string